nexus/src/db/datastore/virtual_provisioning_collection.rs

```diff
@@ -0,0 +1,434 @@
+            disk_byte_diff,
+        )
+        .await
+    }
+
+    pub async fn virtual_provisioning_collection_delete_snapshot(
+        &self,
+        opctx: &OpContext,
+        id: Uuid,
+        project_id: Uuid,
+        disk_byte_diff: i64,
+    ) -> Result<Vec<VirtualProvisioningCollection>, Error> {
+        self.virtual_provisioning_collection_delete_storage(
+            opctx,
+            id,
+            project_id,
+            disk_byte_diff,
+        )
+        .await
+    }
+
+    // Transitively updates all provisioned disk provisions from project -> fleet.
+    async fn virtual_provisioning_collection_delete_storage(
+        &self,
+        opctx: &OpContext,
+        id: Uuid,
+        project_id: Uuid,
+        disk_byte_diff: i64,
+    ) -> Result<Vec<VirtualProvisioningCollection>, Error> {
+        let provisions =
+            VirtualProvisioningCollectionUpdate::new_delete_storage(
+                id,
+                disk_byte_diff,
+                project_id,
+            )
+            .get_results_async(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(e, ErrorHandler::Server)
+            })?;
+        self.virtual_provisioning_collection_producer
+            .append_disk_metrics(&provisions);
+        Ok(provisions)
+    }
+
+    /// Transitively updates all CPU/RAM provisions from project -> fleet.
+    pub async fn virtual_provisioning_collection_insert_instance(
+        &self,
+        opctx: &OpContext,
+        id: Uuid,
+        project_id: Uuid,
+        cpus_diff: i64,
+        ram_diff: i64,
+    ) -> Result<Vec<VirtualProvisioningCollection>, Error> {
+        let provisions =
+            VirtualProvisioningCollectionUpdate::new_insert_instance(
+                id, cpus_diff, ram_diff, project_id,
+            )
+            .get_results_async(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(e, ErrorHandler::Server)
+            })?;
+        self.virtual_provisioning_collection_producer
+            .append_cpu_metrics(&provisions);
+        Ok(provisions)
+    }
+
+    /// Transitively updates all CPU/RAM provisions from project -> fleet.
+    pub async fn virtual_provisioning_collection_delete_instance(
+        &self,
+        opctx: &OpContext,
+        id: Uuid,
+        project_id: Uuid,
+        cpus_diff: i64,
+        ram_diff: i64,
+    ) -> Result<Vec<VirtualProvisioningCollection>, Error> {
+        let provisions =
+            VirtualProvisioningCollectionUpdate::new_delete_instance(
+                id, cpus_diff, ram_diff, project_id,
+            )
+            .get_results_async(self.pool_authorized(opctx).await?)
+            .await
+            .map_err(|e| {
+                public_error_from_diesel_pool(e, ErrorHandler::Server)
+            })?;
+        self.virtual_provisioning_collection_producer
```

Comment:
Sorry if I asked this before -- I went back and looked -- but how does Clickhouse learn the correct values if we crash, say, at this point?

There was also still an open question before about how Clickhouse gets the correct value for, say, a Project, if multiple Nexus got requests to create Instances.

Reply:
Right now, if Nexus crashes, this information might not make it to Clickhouse. Candidly, I'm not sure we can do much about this without significantly changing the structure of the metrics system - at the end of the day, Nexus is a  **producer** of metrics, so even if we put this in a saga, all we can do is place the metrics in an in-memory buffer, and wait for a **collector** (the Oximeter service) to come along, query the metrics, and place them in Clickhouse.

So, how bad is this?
- The view of provisioned resources within **CRDB** is still accurate, even though Clickhouse could be out-of-date. This means that for non-historical usage (e.g., a future implementation of quotas/reservations), we would maintain correctness.
- Since the updates to Clickhouse are "absolute values", not "relative values", the next update to the project would correct the value within Clickhouse. This basically means that the historical view of usage would have a small gap when Nexus is offline, where the un-flushed usage information would be lost.

To mitigate, Nexus could hypothetically make a call to Oximeter, forcing it to make a collection, but this seems like it would basically be a "cache flush" for all metrics whenever any provisioning information changes.

Reply:
I still don't follow how this works with multiple Nexus instances.  To be specific, say:

- Nexus 1 handles request to create Project P
- Nexus 1 handles request to create Instance I1
- Nexus 2 handles request to create Instance I2

Presumably after this, Clickhouse should reflect that P is using the sum of resources used by both I1 and I2.  I assume that means Nexus 2 needs to learn about the resources used by I1 even though it wasn't involved in the provision of I1.  How does that happen?

Reply:
The value being sent to Clickhouse is an *absolute* value, which is only a reflection of "whatever is the latest thing we have stored in CRDB". The value in CRDB is the source-of-truth, which is coordinated between multiple Nexus instances.

- Nexus 1 handles request to create Project P
  - Nexus creates a `virtual_provisioning_collection` CRDB record for the project, and sets it to "zero".
- Nexus 1 handles request to create Instance I1
  - Nexus creates a `virtual_provisioning_resource` CRDB record for `I1`, and simultaneously updates all the parent collections of the resource (which includes `P`) while inserting the record for `I1` itself.
  - This will take the resource usage for `P` from "0" --> "Whatever was used by `I1`"
  - The updated values for all collections are emitted as part of this operation, and can be sent to Clickhouse.
- Nexus 2 handles request to create Instance I2
  - Nexus creates a `virtual_provisioning_resource` CRDB record for `I2`, and simultaneously updates all the parent collections of the resource (which includes `P`) while inserting the record for `I2` itself.
  - This will take the resource usage for `P` from "Whatever was used by `I1`" --> "`I1` + `I2`"
  - The updated values for all collections are emitted as part of this operation, and can be sent to Clickhouse.

Reply:
Thanks, that's helpful.  I think the bit I had missed even after reading this is that the way the database update works [updates the value relatively (as one would expect)](https://github.com/oxidecomputer/omicron/blob/2ab2a3f5f0db9d1c252eee22bbe968f6528a3d63/nexus/src/db/queries/virtual_provisioning_collection_update.rs#L389-L392) and then [fetches the updated [absolute] value](https://github.com/oxidecomputer/omicron/blob/2ab2a3f5f0db9d1c252eee22bbe968f6528a3d63/nexus/src/db/queries/virtual_provisioning_collection_update.rs#L189).  This makes sense.  To answer my question pedantically: Nexus 2 learns about the resources allocated from P for I1 when it makes the database query to update the resources allocated from P for I2 and (as part of the same query) fetches the _total_ resources allocated for P.

(I think it'd be worthwhile to work through this example in a block comment, but definitely not a blocker here.)

The follow-up question then is: what happens if the Clickhouse updates are processed out of order?  Take the same example of two Nexus instances, P, I1, and I2.  From the start: I1 is provisioned from Nexus 1, then I2 is immediately provisioned from Nexus 2.  At this point, say Oximeter hasn't yet collected data from either of them.  Nexus 1 has a pending update for Clickhouse that only includes I1, while Nexus 2 has a (newer) pending update for Clickhouse that includes both I1 and I2.  Now suppose that Oximeter collects from Nexus 2, stores that in Clickhouse, then collects from Nexus 1.  How do we prevent the older value clobbering the correct value?

Reply:
> The follow-up question then is: what happens if the Clickhouse updates are processed out of order?

I took a closer look at this.  I think this is probably okay for MVP but that we may need something more rigorous eventually.  Correct me where this is wrong!

* The underlying Oximeter `Datum` for the Oximeter `Metric`s (`nexus::app::provisioning::CpusProvisioned` etc.) are `i64`, which means they're treated as a gauge (which sounds right).
* That means many values may wind up stored in Clickhouse and exactly one would be used as the "correct" value.  (This is probably obvious but I wanted to check because if this were implemented as, say, counters for bytes-allocated and bytes-freed, the behavior might look different.)
* The _actual_ correct value is whatever was most recently written to CockroachDB.  As far as I can see here, neither Nexus, Oximeter, nor Clickhouse can know this per se.
* Instead, _at the point in the saga where we update the accounting_, we wind up creating an Oximeter `Sample`, which will wind up creating a `Measurement` using `oximeter::Measurement::new()`, which will grab the current wall clock timestamp.  This timestamp will be associated with the measurement in Clickhouse.

As far as I can tell this timestamp is the only thing that can provide an ordering to the various samples that wind up in Clickhouse.  So this process will work as long as the timestamps recorded during the sagas have the same order as the transactions applied at the database.  That's not at all a guarantee (first because system clocks may not be in sync and second because even with sync'd clocks, the transactions could commit in one order while the clients could record their timestamp in another order).  Given CockroachDB's clock requirements I imagine we'll ensure that sled clocks are pretty close.  But with a workload like Terraform that might issue many concurrent provisions, and particularly for a high-level collection like `Silo` or `Fleet` that will be updated for all the activity in the system, a window of even few hundred milliseconds is pretty big.

To be really explicit: the case I'm worried about is that a customer reports that the statistics reported by Clickhouse (well, by our API from the data in Clickhouse) are wrong after somebody runs Terraform.  This could happen even if we keep the clocks pretty tightly sync'd and only have one Nexus instance.  All it would take is for the saga actions to interleave such that the last one to commit to CockroachDB is not one that grabbed the last timestamp.  As far as I can tell, nothing would ever fix the data in Clickhouse until/unless another operation happens that would update these stats.

I can imagine a lot of ways to fix this, none of them trivial.  Some ideas:

- Instead of generating samples for each update, have the producer emit a list of all of the metrics with every collection (i.e., an Oximeter collection winds up being a table scan over the virtual_resource_collection table).  As discussed this can get expensive as we scale up.  But I think this operation isn't latency-sensitive and we don't need to do it that often, right?  We may get pretty far with that approach.
- Put a generation number rather than a timestamp into the Sample and, on query, fetch the one with the latest generation number.  It's appealing because, well, it always seems sketchy to rely on timestamp ordering for correctness -- generation numbers better reflect the truth here.  But it's pretty pedantic and I suspect pretty annoying for people _using_ the metric system.
- Fetch the timestamp from the database so that it's strongly consistent.  We could even put it into the virtual_resource_collection table -- i.e., `time_updated`.  And we'd updated it to `NOW()` in the CTE that updates the row, and we'd fetch it when we fetch back all the columns.  This is conceptually simple and lets metrics consumers just select the latest timestamp (which is a pretty big win I think), but from what I can tell it's much more work than it sounds to plumb that timestamp all the way to the Oximeter `Sample`.

Note that the first option also fixes the case where Nexus crashes.  The others don't.

tl;dr: ship it, file an issue on the above?

Reply:
I spent a while staring at this problem, because:

1. On the one hand, I totally agree with you, it's real and would result in a bug to a customer (portrayed as an inaccurate ordering of values in the metrics graph), and
2. It's feels really goofy to be unable to emit time-based metrics to a timeseries database without hitting this problem. "Record the last value we put in Cockroachdb" *felt* pretty reasonable, and seems like something we might want to do again.

> - Instead of generating samples for each update, have the producer emit a list of all of the metrics with every collection (i.e., an Oximeter collection winds up being a table scan over the virtual_resource_collection table). As discussed this can get expensive as we scale up. But I think this operation isn't latency-sensitive and we don't need to do it that often, right? We may get pretty far with that approach.

I hear you that this is a plausible approach, but it feels like the more we remove "mandatory full-table scans", the better.

> - Put a generation number rather than a timestamp into the Sample and, on query, fetch the one with the latest generation number. It's appealing because, well, it always seems sketchy to rely on timestamp ordering for correctness -- generation numbers better reflect the truth here. But it's pretty pedantic and I suspect pretty annoying for people using the metric system.

Yeah, so this would rely on the consumers of the metrics system doing this re-ordering themselves, which would be especially odd because the samples are already labelled with timestamps (which would tell a different story). Adding a generation number telling a different value would IMO be confusing.

> - Fetch the timestamp from the database so that it's strongly consistent. We could even put it into the virtual_resource_collection table -- i.e., time_updated. And we'd updated it to NOW() in the CTE that updates the row, and we'd fetch it when we fetch back all the columns. This is conceptually simple and lets metrics consumers just select the latest timestamp (which is a pretty big win I think), but from what I can tell it's much more work than it sounds to plumb that timestamp all the way to the Oximeter Sample.

So, this required some work:
- It required enabling a default value for `time_modified` on the insert path, and ensuring we always update it when modifying values. It's critical to always use the clock in CRDB for this value, rather than the one in Nexus.
- It required modifying Oximeter to allow caller-supplied timestamps for samples.

This is done in f96cb955deb8f5c3ac76ea95a75fd4ce7a61d8ca